### PR TITLE
Add upload probing

### PR DESCRIPTION
**Pull request by Pradhyumna02**
Update uploads to use probing data similar to downloads and also wire up these flows to run upload probing to perform upload test

### CODE REVIEW

File: modules/uploadData.js [hunk 1]
```diff
@@ -0,0 +1,1115 @@
...
+    ['100mb', 22876156, 2068],
+    ['100mb', 25163772, 2287],
+    ['100mb', 27680149, 2513],
+    ['100mb', 30448164, 2735]
+];
+
+var bandwidthResults = [];
+
+module.exports = {
+    GetUploadSize: function (bufferSize, time) {
```
Comment (rotosonic) [inline modules/uploadData.js]: Add comments to functions

File: public/lib/uploadHttpConcurrent.js [hunk 2]
```diff
@@ -121,12 +121,12 @@
             if (!this._activeTests.length && this._running) {
                 this.calculateStats();
                 //check to whether increase file size
-                if (this._progressCount > 10) {
-                    //file size can be increased to only certain value as browser is going to crash
-                    if (this.uploadSize < this._maxUploadSize) {
-                        this.uploadSize += this.uploadSize;
-                    }
-                }
+                //if (this._progressCount > 10) {
```
Comment (rotosonic) [inline public/lib/uploadHttpConcurrent.js]: remove commented code

File: public/lib/uploadProbeTest.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            if (typeof(this._activeTests[i]) !== 'undefined') {
+                this._activeTests[i].xhr._request.abort();
+            }
+        }
+    };
+
+
+    window.uploadProbeTest = uploadProbeTest;
+
+})();
```
Comment (rotosonic) [inline public/lib/uploadProbeTest.js]: new line

File: controllers/UploadProbeController.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        this.router.get('/uploadProbe', this.getUPloadProbe.bind(this));
+    }
+
+    /**
+     * endpoint to return response.
+     * @param {request} req - http request.
+     * @param {response} res - http response.
+     */
+    getUPloadProbe(req, res) {
+        console.log(req.query.bufferSize);
```
Comment (kemarc) [inline controllers/UploadProbeController.js]: please remove console.log's

**Comment by kemarc** [comment]
+1, provisional. please remove console.logs

File: public/lib/uploadHttpConcurrentProgress.js [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    uploadHttpConcurrentProgress.prototype.onTestAbort = function () {
+        if (this._running) {
+            if ((Date.now() - this._beginTime) > this.testLength) {
+                if (this._finalResults && this._finalResults.length) {
+                    this.clientCallbackComplete(this._finalResults);
+                } else {
+                    this.clientCallbackError('no measurements obtained');
+                }
+            }
+            this._running=false;
```
Comment (rotosonic) [inline public/lib/uploadHttpConcurrentProgress.js]: Needs to be moved to inside the if statement on line 86

File: public/lib/uploadHttpConcurrentProgress.js [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+            }
+        }
+        else {
+            for (var p = 1; p <= this.concurrentRuns; p++) {
+                this._testIndex++;
+                this['arrayResults' + this._testIndex] = [];
+                this._progressResults['arrayProgressResults' + this._testIndex] = [];
+                request = new window.xmlHttpRequest('POST', this.url, this.timeout, this.onTestComplete.bind(this), this.onTestProgress.bind(this),
+                    this.onTestAbort.bind(this), this.onTestTimeout.bind(this), this.onTestError.bind(this));
+                this._activeTests.push({
```
Comment (rotosonic) [inline public/lib/uploadHttpConcurrentProgress.js]: If the uploads bandwidth test creates multiple http post requests then we would want to create the payload once in this file and pass in the payload to the xmlhttprequest object instead or recreating for each request

File: public/lib/uploadHttpConcurrentProgress.js [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+
+        }
+        this.clientCallbackProgress(totalMovingAverage);
+        this._finalResults.push(totalMovingAverage);
+    };
+
+    /**
+     * onProgress method
+     */
+    uploadHttpConcurrentProgress.prototype.onTestProgress = function (result) { // jshint ignore:line
```
Comment (rotosonic) [inline public/lib/uploadHttpConcurrentProgress.js]: remove //jshing ignore:line

**Comment by rotosonic** [comment]
1+

**Comment by kemarc** [comment]
+1

**Pull request merged**